qualcommax: ipq50xx: add support for Linksys SPNMX56

georgemoussalem opened the pull request:
The SPNMX56 is an ISP-branded and distributed device similar to the MX5500 with the same Wifi chips (IPQ5018 for 2.4G and QCN9074 for 5G) but has an additional QCA8081 PHY providing a 2.5gbps ethernet WAN port.

Speficiations:
* SoC: Qualcomm IPQ5018 (64-bit dual-core ARM Cortex-A53 @ 1.0Ghz)
* Memory: Winbond W634GU6NB-11 (512 MiB DDR3-933)
* Serial Port: 3v3 TTL 115200n8
* Wi-Fi:
	* IPQ5018 (2x2 2.4 Ghz 802.11b/g/n/ax)
	* QCN9024 (4x4:4 5 Ghz 802.11an/ac/ax)
* Ethernet: IPQ5018 integrated virtual switch connected to an 
  external QCA8337 switch (3 Ports 10/100/1000 GBASE-T) and a QCA8081 phy (1 Port up to 2.5 Gbps)
* Flash: Gigadevice GD5F2GM7RExxG (256 MiB)
* LEDs: 1x multi-color PWM LED
* Buttons: 
	* 1x WPS (GPIO 27 Active Low)
	* 1x Reset (GPIO 28 Acive Low)

Flash instructions:
1. On OEM firmware, login to the device (typically at http://192.168.1.1) and click 'CA' in the bottom right corner -> Connectivity -> Manual Upgrade. Alternatively, browse to http://{router IP}/fwupdate.html
Upload openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin 
Optionally flash 2nd partition, after first boot check actual partition:
`fw_printenv -n boot_part`
and install firmware on second partition using command in case of 2:
`mtd -r -e kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin kernel`
and in case of 1:
`mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin alt_kernel`
2. Installation using serial connection from OEM firmware 
hit Enter once booted and enter credentials (login: root, password: admin) 
`fw_printenv -n boot_part`
In case of 2:
```
flash_erase /dev/mtd12 0 0
nandwrite -p /dev/mtd12 openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin
```
or in case of 1:
```
flash_erase /dev/mtd14 0 0
nandwrite -p /dev/mtd14 openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin
```
After first boot install firmware on second partition:
`mtd -r -e kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin kernel`
or:
`mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin alt_kernel`
3. Back to the OEM firmware. Download firmware from OEM website:
Firmware for this device cannot be searched for on the Linksys website. 
Instead, we'd have to use serial to intercept the URL of the firmware while it's trying to update. 
Firmware is ISP specific: 
Toob (UK): http://download.linksys.com/updates/20241125t080737/FW_MX56TB_1.0.1.216218_prod.img

The intention is to collect URLs for different ISPs on a wiki page.

From serial or SSH:
fw_printenv boot_part
in case of 1:
`mtd -r -e alt_kernel -n write FW_MX56TB_1.0.1.216218_prod.img alt_kernel`
else in case of 2:
`mtd -r -e kernel -n write FW_MX56TB_1.0.1.216218_prod.img kernel`

georgemoussalem committed:
qualcommax: ipq50xx: add support for Linksys SPNMX56

The SPNMX56 is an ISP-branded and distributed device similar to the MX5500
with the same Wifi chips (IPQ5018 for 2.4G and QCN9074 for 5G) but has an
additional QCA8081 PHY providing a 2.5gbps ethernet WAN port.

Speficiations:
* SoC: Qualcomm IPQ5018 (64-bit dual-core ARM Cortex-A53 @ 1.0Ghz)
* Memory: Winbond W634GU6NB-11 (512 MiB DDR3-933)
* Serial Port: 3v3 TTL 115200n8
* Wi-Fi: IPQ5018 (2x2 2.4 Ghz 802.11b/g/n/ax)
         QCN9024 (4x4:4 5 Ghz 802.11an/ac/ax)
* Ethernet: IPQ5018 integrated virtual switch connected to an external
            QCA8337 switch (3 Ports 10/100/1000 GBASE-T) and a
            QCA8081 phy (up to 2.5 Gbps)
* Flash: Gigadevice GD5F2GM7RExxG (256 MiB)
* LEDs: 1x multi-color PWM LED
* Buttons: 1x WPS (GPIO 27 Active Low)
           1x Reset (GPIO 28 Acive Low)

Flash instructions:
1. On OEM firmware, login to the device (typically at http://192.168.1.1)
and click 'CA' in the bottom right corner -> Connectivity ->
Manual Upgrade. Alternatively, browse to http://<router IP>/fwupdate.html
Upload openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin
Optionally flash 2nd partition, after first boot check actual partition:
fw_printenv -n boot_part
and install firmware on second partition using command in case of 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin kernel
and in case of 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin alt_kernel
2. Installation using serial connection from OEM firmware
hit Enter once booted and enter credentials (login: root, password: admin)
fw_printenv -n boot_part
In case of 2:
flash_erase /dev/mtd12 0 0
nandwrite -p /dev/mtd12 openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin
or in case of 1:
flash_erase /dev/mtd14 0 0
nandwrite -p /dev/mtd14 openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin
After first boot install firmware on second partition:
mtd -r -e kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin kernel
or:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq50xx-linksys_spnmx56-squashfs-factory.bin alt_kernel
3. Back to the OEM firmware.
Download firmware from OEM website:
Firmware for this device cannot be searched for on the Linksys website.
Instead, we'd have to use serial to intercept the URL of the firmware
while it's trying to update. Firmware is ISP specific:
Toob (UK): http://download.linksys.com/updates/20241125t080737/FW_MX56TB_1.0.1.216218_prod.img

The intention is to collect URLs for different ISPs on a wiki page.

From serial or SSH:
fw_printenv boot_part
in case of 1:
mtd -r -e alt_kernel -n write FW_MX56TB_1.0.1.216218_prod.img alt_kernel
else in case of 2:
mtd -r -e kernel -n write FW_MX56TB_1.0.1.216218_prod.img kernel

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/17968
Signed-off-by: Robert Marko <robimarko@gmail.com>

testuser7 commented on target/linux/qualcommax/ipq50xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -73,7 +73,8 @@ platform_check_image() {
-	linksys,mx5500)
+	linksys,mx5500|\
+	linksys,spnmx56)
 		remove_oem_ubi_volume rootfs
```
Volume name that should be removed is not correct: https://github.com/openwrt/openwrt/commit/398f4a97378e2f645badc1aef3d0e9fd76f6665d#commitcomment-152522760

georgemoussalem replied:
ah, thanks! I took the name from:
```root@ae-ap08:~# cat /sys/class/ubi/ubi0/ubi0_0/name
rootfs``` 
in openwrt..

let me use this PR to add a commit to the name for mx2000, mx5500, and spnmx56 at once.

georgemoussalem committed:
qualcommax: ipq50xx: sysupgrade: fix OEM UBI volume name for Linksys MX devices

The UBI volume name in OEM firmware for Linksys MX2000, MX5500, and
SPNMX56 is 'squashfs' instead of 'rootfs'. So let's update the volume
accordingly.

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/17968
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Please update ipq-wifi to point to a version that includes the required BDF-s (Use the maintainer script: https://git.openwrt.org/?p=maintainer-tools.git;a=blob;f=update_git_source_package.sh)

georgemoussalem commented:
> Please update ipq-wifi to point to a version that includes the required BDF-s (Use the maintainer script: https://git.openwrt.org/?p=maintainer-tools.git;a=blob;f=update_git_source_package.sh)

done, ready to go

georgemoussalem commented:
> the commit created by the maintainer script is causing the formalities to fail now due to an empty commit message. Any action on me?

Rebased on top of main and included all the commits up to the commit for SPNMX56. 
CI test formalities shouldn't complain now. Anything else needed for merge @robimarko?

georgemoussalem commented:
> You can now drop the ipq-wifi, I merged it as a part of another PR.

done, removed the commit for ipq-wifi

robimarko commented:
Thanks! Rebased on top of main and merged!

Daviey commented:
@georgemoussalem thanks for adding this support, I can confirm that I've been able to flash 2 x boxes with the latest snapshot.

However, the WAN port is supposed to support 2.5GbE but this doesn't work.  When trying to autoneg to this, the interface fails to come up (seeing no response to DHCP).

Autoneg to a network supporting only 100 works as expected, and using ethtool to disable autoneg and peg to 100 also works with a 2.5GbE network.

Is this the same behaviour you are seeing?  Thanks
